### ISSUE TITLE
Clang allows `nullptr` dereference in `constexpr`

### ISSUE THREAD

**Issue by poconn** [llvm/llvm-project#69108]
Hi! Clang accepts dereferencing a null pointer in a `constexpr` context when binding a reference to it but not reading the contents:

```cpp
constexpr auto b = [] {
  int* px = nullptr;
  int& x = *px;
  return true;
}();
```

MSVC and GCC both reject it:
https://godbolt.org/z/frceTafjj

**Comment by shafik** [comment]
So it looks like clang allows binding to a reference but not actual dereference: https://godbolt.org/z/13PaGs3Wj

```cpp
bool constexpr f() {
  int* px = nullptr;
  int x = *px;
  return true;
}

bool constexpr f2() {
  int* px = nullptr;
  int &x = *px;
  return x;
}

void g() {
    constexpr bool b
      = f();
    constexpr bool b2
      = f2();
}

```

I think binding to a reference should be ok CC @zygoloid

**Comment by RungeCC** [comment]
According to https://cplusplus.github.io/CWG/issues/315.html (NAD) I think the original case is well-formed.

Since:

1. builtin indirection expression yield an lvalue ([expr.unary.op]) and
2. binding an lvalue reference of int to an lvalue expression of type int does not need to perform lvalue-to-rvalue conversion ([dcl.init.ref]).

**Comment by shafik** [comment]
Apologies, the intention should be that this is undefined behavior [dcl.ref p5](https://eel.is/c++draft/dcl.ref#5):

> A reference shall be initialized to refer to a valid object or function[.](https://eel.is/c++draft/dcl.ref#5.sentence-3)

and

> In particular, a null reference cannot exist in a well-defined program, because the only way to create such a reference would be to bind it to the “object” obtained by indirection through a null pointer, which causes undefined behavior

See this is diagnosed by clang: https://godbolt.org/z/j8M44nTeq

```cpp
void f() {
    constexpr int &y = *(int*)nullptr;
    constexpr int &x = x;
}
```

The normative wording could be clarified though see [cwg453](https://wg21.link/cwg453)

**Comment by zygoloid** [comment]
Yes, it's OK to form an lvalue that doesn't denote an object (an "empty lvalue") but not to bind a reference to such an lvalue.

**Comment by shafik** [comment]
We diagnose it in this case but the diagnostic is not great: https://godbolt.org/z/qMTT35c1q

```cpp
void f() {
  constexpr int const* px = nullptr;
  constexpr int const &x = *px;
}
```

**Comment by Sunday111** [comment]
Doesn't reproduce in 22.1.0: https://godbolt.org/z/bavxe1sve
